File: tests/go/generics.go
// Package generics demonstrates Go 1.18+ generic types and functions.
// It covers type parameters, constraints, multi-line signatures, and
// methods on generic types.
package generics

import "cmp"

// Ordered is a constraint that permits any ordered type.
type Ordered interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64 |
		~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr |
		~float32 | ~float64 | ~string
}

// Pair holds two values of potentially different types.
type Pair[A, B any] struct {
	First  A
	Second B
}

// NewPair constructs a Pair.
func NewPair[A, B any](first A, second B) Pair[A, B] {
	return Pair[A, B]{First: first, Second: second}
}

func (p Pair[A, B]) Swap() Pair[B, A] {
	return Pair[B, A]{}
}

// Stack is a generic LIFO container.
type Stack[T any] struct {
	items []T
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (s *Stack[T]) Push(item T) {
	// body elided
}

func (s *Stack[T]) Pop() (T, bool) {
	var zero T
	return zero, false
}

func (s *Stack[T]) Peek() (T, bool) {
	var zero T
	return zero, false
}

func (s *Stack[T]) Len() int {
	return 0
}

// Map applies a transform function to each element.
func Map[T, U any](slice []T, f func(T) U) []U {
	return nil
}

// Filter retains only elements satisfying the predicate.
func Filter[T any](slice []T, pred func(T) bool) []T {
	return nil
}

// Reduce folds a slice into a single value.
func Reduce[T, U any](slice []T, init U, f func(U, T) U) U {
	return init
}

// Min returns the smaller of two ordered values.
func Min[T cmp.Ordered](a, b T) T {
	if a < b {
		return a
	}
	return b
}

// Max returns the larger of two ordered values.
func Max[T cmp.Ordered](a, b T) T {
	if a > b {
		return a
	}
	return b
}

// Set is a generic hash set backed by a map.
type Set[T comparable] struct {
	m map[T]struct{}
}

func NewSet[T comparable]() *Set[T] {
	return &Set[T]{m: make(map[T]struct{})}
}

func (s *Set[T]) Add(v T) {
	// body elided
}

func (s *Set[T]) Contains(v T) bool {
	return false
}

func (s *Set[T]) Remove(v T) {
	// body elided
}

func (s *Set[T]) Len() int {
	return 0
}

// Result wraps a value or an error, similar to Rust's Result.
type Result[T any] struct {
	value T
	err   error
}

func Ok[T any](v T) Result[T] {
	return Result[T]{value: v}
}

func Err[T any](err error) Result[T] {
	return Result[T]{err: err}
}

func (r Result[T]) Unwrap() T {
	return r.value
}

func (r Result[T]) IsOk() bool {
	return r.err == nil
}

File: tests/kotlin/generics.kt
package com.example.generics

const val DEFAULT_CAPACITY = 16

interface Container<T> {
    fun add(item: T): Boolean
    fun remove(item: T): Boolean
    fun contains(item: T): Boolean
    fun size(): Int
    fun isEmpty(): Boolean
}

interface Transformer<in A, out B> {
    fun transform(input: A): B
}

interface Comparable2<T> {
    operator fun compareTo(other: T): Int
}

data class Pair2<A, B>(val first: A, val second: B) {
    fun swap(): Pair2<B, A> = Pair2(second, first)
    fun <C> mapFirst(f: (A) -> C): Pair2<C, B> = Pair2(f(first), second)
}

class BoundedList<T : Comparable<T>>(private val capacity: Int = DEFAULT_CAPACITY) : Container<T> {
    private val items = mutableListOf<T>()

    override fun add(item: T): Boolean {
        if (items.size >= capacity) return false
        return items.add(item)
    }

    override fun remove(item: T): Boolean = items.remove(item)

    override fun contains(item: T): Boolean = item in items

    override fun size(): Int = items.size

    override fun isEmpty(): Boolean = items.isEmpty()

    fun min(): T? = items.minOrNull()
    fun max(): T? = items.maxOrNull()

    fun <R : Comparable<R>> sortedBy(selector: (T) -> R): List<T> {
        return items.sortedBy(selector)
    }
}

sealed class Result<out T, out E> {
    data class Ok<T>(val value: T) : Result<T, Nothing>()
    data class Err<E>(val error: E) : Result<Nothing, E>()

    fun isOk(): Boolean = this is Ok
    fun isErr(): Boolean = this is Err

    fun <U> map(transform: (T) -> U): Result<U, E> {
        return when (this) {
            is Ok -> Ok(transform(value))
            is Err -> this
        }
    }
}

inline fun <T, reified R> List<T>.filterIsInstanceAndMap(transform: (R) -> R): List<R> {
    return filterIsInstance<R>().map(transform)
}

fun <T> List<T>.second(): T? = if (size >= 2) get(1) else null

fun <K, V> Map<K, V>.getOrThrow(key: K): V {
    return get(key) ?: throw NoSuchElementException("Key $key not found")
}

fun <T : Any> T?.orThrow(message: String = "Expected non-null value"): T {
    return this ?: throw NullPointerException(message)
}

fun <A, B, C> compose(f: (B) -> C, g: (A) -> B): (A) -> C {
    return { a -> f(g(a)) }
}

object TypeUtils {
    inline fun <reified T> isType(value: Any): Boolean = value is T
    fun <T> identity(value: T): T = value
    fun <T, R> List<T>.mapNotNullTo(transform: (T) -> R?): List<R> {
        return mapNotNull(transform)
    }
}
